### PR TITLE
reed solomon perf improvements proposal

### PR DESCRIPTION
**Pull request by drahnr**
@burdges @TomH149

### CODE REVIEW

**Comment by burdges** [comment]
LGTM

File: applications/reed-solomon-perf.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+Improve reed solomon erasure coding used in polkadot to reduce the static overhead the current implementation.
+
+
+
+### Project Details
+
+We expect the teams to already have a solid idea about your project's expected final state. Therefore, we ask the teams to submit (where relevant):
+
+* introduce SMID instructions to the current implementation to improve performance for 1000+ validators
```
Comment (ordian) [inline applications/reed-solomon-perf.md:29]:
```suggestion
* introduce SIMD instructions to the current implementation to improve performance for 1000+ validators
```

File: applications/reed-solomon-perf.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+### Team LinkedIn Profiles (if available)
+
+- https://www.linkedin.com/in/bernhardschuster
+
+## Development Roadmap :nut_and_bolt:
+
+1. do performance measurements baseline
+2. setup decommissioned CI infra again
+3. add SMID to the current implementation
```
Comment (ordian) [inline applications/reed-solomon-perf.md:77]:
```suggestion
3. add SIMD to the current implementation
```

File: applications/reed-solomon-perf.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+- **Estimated duration:** 1 month
+- **FTE:**  1
+- **Costs:** 10,560 EUR
+
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| **0a.** | License | Apache 2.0 / MIT |
+| **0b.** | Documentation | We will provide both **inline documentation** of the code and an outline |
+| **0c.** | Testing and Testing Guide | Core functions will be fully covered by comprehensive unit tests to ensure functionality and robustness. |
+| 1. | SMID perf | Implement SMID |
```
Comment (ordian) [inline applications/reed-solomon-perf.md:102]:
```suggestion
| 1. | SIMD perf | Implement SIMD |
```

**Review by ordian — COMMENTED** [review]
Hey Bernhard :)

When exploring SIMD, would be great to do runtime feature detection, like in https://github.com/dalek-cryptography/curve25519-dalek/pull/523

**Comment by ordian** [comment]
Also relevant: https://github.com/paritytech/polkadot/issues/7505

**Comment by burdges** [comment]
> Also relevant: https://github.com/paritytech/polkadot/issues/7505

This is internal to how polkadot uses the chunks.  I don't care who does this work of course.  And the migration is a simply an if check to start the shuffling at some fixed block height, plus something to deprecate hosts who won't have this check.  I'll send some emails about my generalized Feistel algorithm question for that.

**Comment by ordian** [comment]
I meant making the encoding systematic, which is currently not on the https://github.com/paritytech/reed-solomon-novelpoly/ level IIUC. @drahnr please correct me if I'm wrong

**Comment by burdges** [comment]
I'd thought it was systematic from the beginning, but if not then yeah reevaluate that choice.

**Comment by drahnr** [comment]
> Hey Bernhard :)
> 
> When exploring SIMD, would be great to do runtime feature detection, like in [dalek-cryptography/curve25519-dalek#523](https://github.com/dalek-cryptography/curve25519-dalek/pull/523)

Will do!

**Comment by drahnr** [comment]
> I'd thought it was systematic from the beginning, but if not then yeah we should recall why we made that choice, but that'd be larger than this proposal.

From the paper, it should be, I didn't verify the impl yet

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application. I'm generally happy to go ahead with it. Just a few quick comments: 
- Feel free to remove the template text that isn't bold from the application, like "We expect the teams to already have a solid idea about your project's expected final state. Therefore, we ask the teams to submit (where relevant)". This makes it hard to read.
- Could you add the programming language to the deliverables? 
- Also feel free to add 

> When exploring SIMD, would be great to do runtime feature detection, like in https://github.com/dalek-cryptography/curve25519-dalek/pull/523

**Review by ordian — APPROVED** [review]

**Review by Noc2 — APPROVED** [review]
Thanks for the update. I'm happy to go ahead with it.

**Review by xlc — APPROVED** [review]

**Comment by riusricardo** [comment]
@drahnr Thanks for the application. After merging, I realized that there's a discrepancy. Could you please clarify the difference in the total amount and the amounts per milestone? The sum doesn't add up.

**Comment by drahnr** [comment]
See PR #3 which addresses the inconsistencies

**Comment by drahnr** [comment]
> > Hey Bernhard :)
> > When exploring SIMD, would be great to do runtime feature detection, like in [dalek-cryptography/curve25519-dalek#523](https://github.com/dalek-cryptography/curve25519-dalek/pull/523)
> 
> Will do, given it doesn't explode time wise

After a first time assessment, it will not be part of the milestone 1 implementation.

**Comment by drahnr** [comment]
A bit of brief update, I am currently pending some clarification on the gao mateer paper by the authors. Otherwise it's well on track to delivery, albeit a bit behind schedule for a bag of personal reasons (good ones only!).

**Comment by ordian** [comment]
jFYI: https://github.com/paritytech/reed-solomon-novelpoly/issues/40, SIMD impl has a 10x speedup over non-SIMD (also in the original C++ leopard lib), so it's worth looking into more closely.

**Comment by drahnr** [comment]
> jFYI: [paritytech/reed-solomon-novelpoly#40](https://github.com/paritytech/reed-solomon-novelpoly/issues/40), SIMD impl has a 10x speedup over non-SIMD (also in the original C++ leopard lib), so it's worth looking into more closely.

Jeff already mentioned it. I'd see that as part of #10 , going to make it explicit there, I don't have capacity to take on unpaid additional work beyond what I do already as open source work unfortunately and part 1 is already closed.

**Comment by drahnr** [comment]
So turns out I underestimated the time required for it. It's still going to be a bit before there are results. Just to be clear, further proposals will come in at a more realistic rate taking into account the uncertainty bits more.

**Pull request merged**